sqltoken/tokenize.go

```diff
@@ -0,0 +1,730 @@
+				goto Dollar
+			}
+			token(Punctuation)
+		case 'U':
+			// U&'d\0061t\+000061'
+			if config.NoticeUAmpPrefix && i+1 < len(s) && s[i] == '&' && s[i+1] == '\'' {
+				i += 2
+				goto SingleQuoteString
+			}
+			goto Word
+		case 'x', 'X':
+			// X'1f' x'1f'
+			if config.NoticeHexNumbers && i < len(s) && s[i] == '\'' {
+				i++
+				goto QuotedHexNumber
+			}
+			goto Word
+		case 'b', 'B':
+			if config.NoticeBinaryNumbers && i < len(s) && s[i] == '\'' {
+				i++
+				goto QuotedBinaryNumber
+			}
+			goto Word
+		case 'a' /*b*/, 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm',
+			'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w' /*x*/, 'y', 'z',
+			'A' /*B*/, 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M',
+			'N', 'O', 'P', 'Q', 'R', 'S', 'T' /*U*/, 'V', 'W' /*X*/, 'Y', 'Z',
+			'_':
+			goto Word
+		case '0':
+			if config.NoticeHexNumbers && i < len(s) && s[i] == 'x' {
+				i++
+				goto HexNumber
+			}
+			if config.NoticeBinaryNumbers && i < len(s) && s[i] == 'b' {
+				i++
+				goto BinaryNumber
+			}
+			goto Number
+		case /*0*/ '1', '2', '3', '4', '5', '6', '7', '8', '9':
+			goto Number
+		default:
+			r, w := utf8.DecodeRuneInString(s[i-1:])
+			switch {
+			case r == '⎖':
+				i += w - 1
+				goto NumberNoDot
+			case unicode.IsDigit(r):
+				i += w - 1
+				goto Number
+			case unicode.IsPunct(r) || unicode.IsSymbol(r) || unicode.IsMark(r):
+				i += w - 1
+				token(Punctuation)
+			case unicode.IsLetter(r):
+				i += w - 1
+				goto Word
+			case unicode.IsControl(r) || unicode.IsSpace(r):
+				i += w - 1
+				goto Whitespace
+			default:
+				i += w - 1
+				token(Other)
+			}
+		}
+	}
+	goto Done
+
+CStyleComment:
+	for i < len(s) {
+		c := s[i]
+		i++
+		switch c {
+		case '*':
+			if i < len(s) && s[i] == '/' {
+				i++
+				token(Comment)
+				goto BaseState
+			}
+		}
+	}
+	token(Comment)
+	goto Done
+
+SingleQuoteString:
+	for i < len(s) {
+		c := s[i]
+		i++
+		switch c {
+		case '\'':
+			token(Literal)
+			goto BaseState
+		case '\\':
+			if i < len(s) {
+				i++
+			} else {
+				token(Literal)
+				goto Done
+			}
+		}
+	}
+	token(Literal)
+	goto Done
+
+DoubleQuoteString:
+	for i < len(s) {
+		c := s[i]
+		i++
+		switch c {
+		case '"':
+			token(Literal)
+			goto BaseState
+		case '\\':
+			if i < len(s) {
+				i++
+			} else {
+				token(Literal)
+				goto Done
+			}
+		}
+	}
+	token(Literal)
+	goto Done
+
+SkipToEOL:
+	for i < len(s) {
+		c := s[i]
+		i++
+		switch c {
+		case '\n':
+			token(Comment)
+			goto BaseState
+		}
+	}
+	token(Comment)
+	goto Done
+
+Word:
+	for i < len(s) {
+		c := s[i]
+		switch c {
+		case 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm',
+			'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z',
+			'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M',
+			'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z',
+			'_',
+			'0', '1', '2', '3', '4', '5', '6', '7', '8', '9':
```

Comment:
I think this would be cleaner with range expressions inside a general `switch {`

Reply:
I've added a comment: this is a performance hack.  Would a range expression perform better?

Reply:
I would think the performance difference with a range expression would be imperceptible in the context of running database migrations.

Reply:
I'm planning to use this code to fix sqlx --- it would be run on nearly every (or maybe every?) query done by sqlx.   Performance matters in that context.